planner/cascades/transformation_rules.go

```diff
@@ -2502,3 +2505,133 @@ func (r *PullSelectionUpApply) OnTransform(old *memo.ExprIter) (newExprs []*memo
 	newApplyGroupExpr.SetChildren(outerChildGroup, old.Children[1].GetExpr().Children[0])
 	return []*memo.GroupExpr{newApplyGroupExpr}, false, false, nil
 }
+
+// MergeAdjacentWindow merge adjacent Window.
+type MergeAdjacentWindow struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentWindow creates a new Transformation MergeAdjacentWindow.
+// The pattern of this rule is `Window -> Window`.
+func NewRuleMergeAdjacentWindow() Transformation {
+	rule := &MergeAdjacentWindow{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandWindow,
+		memo.EngineAll,
+		memo.NewPattern(memo.OperandWindow, memo.EngineAll),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *MergeAdjacentWindow) Match(expr *memo.ExprIter) bool {
+	curWinPlan := expr.GetExpr().ExprNode.(*plannercore.LogicalWindow)
+	nextGroupExpr := expr.Children[0].GetExpr()
+	nextWinPlan := nextGroupExpr.ExprNode.(*plannercore.LogicalWindow)
+	nextSchema := nextGroupExpr.Group.Prop.Schema
+	nextGroupChildren := nextGroupExpr.Children
+	ctx := expr.GetExpr().ExprNode.SCtx()
+
+	// Whether Partition parts are the same.
+	if len(curWinPlan.PartitionBy) != len(nextWinPlan.PartitionBy) ||
+		len(curWinPlan.OrderBy) != len(nextWinPlan.OrderBy) ||
+		(curWinPlan.Frame == nil && nextWinPlan.Frame != nil) ||
+		(curWinPlan.Frame != nil && nextWinPlan.Frame == nil) {
+		return false
+	}
+	for i, item := range curWinPlan.PartitionBy {
+		if !item.Col.Equal(ctx, nextWinPlan.PartitionBy[i].Col) ||
+			item.Desc != nextWinPlan.PartitionBy[i].Desc {
+			return false
```

Comment:
The `partition_clause` has this syntax: `PARTITION BY expr [, expr] ...`, so we don't need to check `Desc` of `PartitionBy`

Is the result of the `partition by a, b` same as `partition by b, a`? 
Should we merge them in this case?

Reply:
> The `partition_clause` has this syntax: `PARTITION BY expr [, expr] ...`, so IMO, we don't need to check `Desc` of `PartitionBy`
> 

> Is the result of the `partition by a, b` same as `partition by b, a`?
> Should we merge them in this case?

@mmyj Maybe you are right. I hold that the columns in partition clause cannot affect the result of window, but the implementation of partititon clause (`property.SortItem`) makes me confused. 

By the way, why parititon clause is implemented by `SortItem`?

Reply:
I'm confused also.

Reply:
> I'm confused also.

@mmyj I remove the order-consistent-force of `PartitionBy` in [`newer commit`](https://github.com/pingcap/tidb/pull/24556/commits/071f638ba9a88c5b87f105b7b5693276defaf37a) just now.